plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcCastPushdownTest.java

```diff
@@ -0,0 +1,94 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.testing.AbstractTestQueryFramework;
+import io.trino.testing.sql.SqlExecutor;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public abstract class BaseJdbcCastPushdownTest
+        extends AbstractTestQueryFramework
+{
+    protected abstract String leftTable();
+
+    protected abstract String rightTable();
+
+    protected abstract SqlExecutor onRemoteDatabase();
+
+    protected abstract List<CastTestCase> supportedCastTypePushdown();
+
+    protected abstract List<CastTestCase> unsupportedCastTypePushdown();
+
+    protected abstract List<CastTestCase> failCast();
+
+    @Test
+    public void testProjectionPushdownWithCast()
+    {
+        for (CastTestCase testCase : supportedCastTypePushdown()) {
+            assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                    .isFullyPushedDown();
+        }
+
+        for (CastTestCase testCase : unsupportedCastTypePushdown()) {
+            assertThat(query("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
+                    .isNotFullyPushedDown(ProjectNode.class);
+        }
+    }
+
+    @Test
+    public void testJoinPushdownWithCast()
+    {
+        for (CastTestCase testCase : supportedCastTypePushdown()) {
+            assertThat(query("SELECT l.id FROM %s l JOIN %s r ON CAST(l.%s AS %s) = r.%s".formatted(leftTable(), rightTable(), testCase.sourceColumn(), testCase.castType(), testCase.targetColumn().orElseThrow())))
+                    .isFullyPushedDown();
+        }
+
+        for (CastTestCase testCase : unsupportedCastTypePushdown()) {
+            assertThat(query("SELECT l.id FROM %s l JOIN %s r ON CAST(l.%s AS %s) = r.%s".formatted(leftTable(), rightTable(), testCase.sourceColumn(), testCase.castType(), testCase.targetColumn().orElseThrow())))
+                    .joinIsNotFullyPushedDown();
+        }
+    }
+
+    @Test
+    public void testCastFails()
+    {
+        for (CastTestCase testCase : failCast()) {
+            assertThatThrownBy(() -> getQueryRunner().execute("SELECT CAST(%s AS %s) FROM %s".formatted(testCase.sourceColumn(), testCase.castType(), leftTable())))
```

Comment:
When this test fails, it doesn't say which condition has failed exactly. Do you think we could improve on that?
Took a quick look - `as()` or `withFailMessage()` does not seem to work, maybe custom `satifies()` would.

Reply:
Thanks @SemionPar, I will take this as a follow up.

Reply:
@krvikash Sure thing!

One more thought that crossed my mind: if cast fails at runtime with different message when pushed down, say:

```
      Trino: Cannot cast DECIMAL(19, 2) '99999999999999999.99' to DECIMAL(10, 2)
      Teradata: [Error 2616] [SQLState 22003] Numeric overflow occurred during computation.
```

It would be good to be able to use `testCastFails` to test such cases, don't you think? We might want to add message customization to `CastTestCase` (or introduce `FailCastTestCase`) - consider this another improvement idea.

Reply:
@SemionPar Here is the refactoring PR https://github.com/trinodb/trino/pull/23737